File: JavaScript/src/Internals/MyDate.js
/**
 * 日期时间处理类
 */
export class MyDate {
    constructor(year, month, day, hour, minute, second) {
        if (year instanceof Date) {
            // 构造函数重载：接受Date对象
            this.Year = year.getFullYear();
            this.Month = year.getMonth() + 1;
            this.Day = year.getDate();
            this.Hour = year.getHours();
            this.Minute = year.getMinutes();
            this.Second = year.getSeconds();
        } else if (typeof year === 'number' && month === undefined) {
            // 构造函数重载：接受数字（像Excel日期值）
            let num = year;
            let days = Math.floor(num);
            if (days > 365) {
                // 从1900年1月1日开始计算
                let start = new Date(1900, 0, 1);
                let targetDate = new Date(start);
                targetDate.setDate(targetDate.getDate() + days - 2);
                this.Year = targetDate.getFullYear();
                this.Month = targetDate.getMonth() + 1;
                this.Day = targetDate.getDate();
            } else {
                this.Year = null;
                this.Month = null;
                this.Day = days;
            }
            let d = num - days;
            this.Hour = Math.floor(d * 24);
            this.Minute = Math.floor((d * 24 - this.Hour) * 60);
            this.Second = Math.round(((d * 24 - this.Hour) * 60 - this.Minute) * 60);
            if (this.Second === 60) {
                this.Second = 0;
                this.Minute++;
                if (this.Minute === 60) {
                    this.Minute = 0;
                    this.Hour++;
                }
            }
        } else if (typeof year === 'object' && year !== null && 'days' in year) {
            // 构造函数重载：接受时间跨度对象
            let timeSpan = year;
            this.Year = null;
            this.Month = null;
            this.Day = timeSpan.days || 0;
            this.Hour = timeSpan.hours || 0;
            this.Minute = timeSpan.minutes || 0;
            this.Second = timeSpan.seconds || 0;
        } else {
            // 构造函数重载：接受年、月、日、时、分、秒
            this.Year = year;
            this.Month = month;
            this.Day = day;
            this.Hour = hour || 0;
            this.Minute = minute || 0;
            this.Second = second || 0;
        }
    }

    /**
     * 转换为字符串
     */
    toString(format) {
        if (format) {
            // 这里可以实现格式化功能，暂时返回默认格式
            return this.toString();
        }
        if (this.Year != null && this.Year > 0) {
            return `${this.Year}-${this.Month.toString().padStart(2, '0')}-${this.Day.toString().padStart(2, '0')} ${this.Hour.toString().padStart(2, '0')}:${this.Minute.toString().padStart(2, '0')}:${this.Second.toString().padStart(2, '0')}`;
        } else {
            let result = '';
            if (this.Day != null && this.Day > 0) {
                result += `${this.Day} `;
            }
            result += `${this.Hour.toString().padStart(2, '0')}:${this.Minute.toString().padStart(2, '0')}:${this.Second.toString().padStart(2, '0')}`;
            return result;
        }
    }

    /**
     * 转换为Date对象
     */
    ToDateTime(dateTimeKind) {
        if (this.Year != null && this.Year > 0) {
            if (dateTimeKind === 0) {
                // UTC时间
                return new Date(Date.UTC(this.Year, this.Month - 1, this.Day, this.Hour, this.Minute, this.Second));
            } else {
                // 本地时间
                return new Date(this.Year, this.Month - 1, this.Day, this.Hour, this.Minute, this.Second);
            }
        }
        // 只有时间部分，返回从1970年1月1日开始的时间
        return new Date(1970, 0, 1 + (this.Day || 0), this.Hour, this.Minute, this.Second);
    }

    /**
     * 转换为时间跨度对象
     */
    ToTimeSpan() {
        return {
            days: this.Day || 0,
            hours: this.Hour,
            minutes: this.Minute,
            seconds: this.Second
        };
    }

    /**
     * 添加年
     */
    AddYears(year) {
        let newYear = (this.Year || 0) + year;
        return new MyDate(newYear, this.Month, this.Day, this.Hour, this.Minute, this.Second);
    }

    /**
     * 添加月
     */
    AddMonths(month) {
        let newMonth = (this.Month || 0) + month;
        if (newMonth >= 1 && newMonth <= 12) {
            return new MyDate(this.Year, newMonth, this.Day, this.Hour, this.Minute, this.Second);
        }
        // 如果月份超出范围，使用Date对象计算
        let date = this.ToDateTime();
        date.setMonth(date.getMonth() + month);
        return new MyDate(date);
    }

    /**
     * 添加日
     */
    AddDays(day) {
        if (this.Year != null && this.Year > 1900) {
            let newDay = (this.Day || 0) + day;
            if (newDay >= 1 && newDay <= 28) {
                return new MyDate(this.Year, this.Month, newDay, this.Hour, this.Minute, this.Second);
            }
            // 如果日期超出范围，使用Date对象计算
            let date = this.ToDateTime();
            date.setDate(date.getDate() + day);
            return new MyDate(date);
        }
        // 只有时间部分，直接添加天数
        return new MyDate(null, null, (this.Day || 0) + day, this.Hour, this.Minute, this.Second);
    }

    /**
     * 添加时
     */
    AddHours(hour) {
        let newHour = this.Hour + hour;
        if (newHour >= 0 && newHour < 24) {
            return new MyDate(this.Year, this.Month, this.Day, newHour, this.Minute, this.Second);
        }
        if (this.Year != null && this.Year > 1900) {
            // 使用Date对象计算
            let date = this.ToDateTime();
            date.setHours(date.getHours() + hour);
            return new MyDate(date);
        }
        // 只有时间部分，计算总小时数
        let totalHours = this.Hour + hour;
        let days = Math.floor(totalHours / 24);
        let remainingHours = totalHours % 24;
        if (remainingHours < 0) {
            remainingHours += 24;
            days--;
        }
        return new MyDate(null, null, (this.Day || 0) + days, remainingHours, this.Minute, this.Second);
    }

    /**
     * 添加分
     */
    AddMinutes(minute) {
        let newMinute = this.Minute + minute;
        if (newMinute >= 0 && newMinute <= 59) {
            return new MyDate(this.Year, this.Month, this.Day, this.Hour, newMinute, this.Second);
        }
        if (this.Year != null && this.Year > 1900) {
            // 使用Date对象计算
            let date = this.ToDateTime();
            date.setMinutes(date.getMinutes() + minute);
            return new MyDate(date);
        }
        // 只有时间部分，计算总分钟数
        let totalMinutes = this.Minute + minute;
        let hours = Math.floor(totalMinutes / 60);
        let remainingMinutes = totalMinutes % 60;
        if (remainingMinutes < 0) {
            remainingMinutes += 60;
            hours--;
        }
        return this.AddHours(hours).AddMinutes(remainingMinutes);
    }

    /**
     * 添加秒
     */
    AddSeconds(second) {
        let newSecond = this.Second + second;
        if (newSecond >= 0 && newSecond <= 59) {
            return new MyDate(this.Year, this.Month, this.Day, this.Hour, this.Minute, newSecond);
        }
        if (this.Year != null && this.Year > 1900) {
            // 使用Date对象计算
            let date = this.ToDateTime();
            date.setSeconds(date.getSeconds() + second);
            return new MyDate(date);
        }
        // 只有时间部分，计算总秒数
        let totalSeconds = this.Second + second;
        let minutes = Math.floor(totalSeconds / 60);
        let remainingSeconds = totalSeconds % 60;
        if (remainingSeconds < 0) {
            remainingSeconds += 60;
            minutes--;
        }
        return this.AddMinutes(minutes).AddSeconds(remainingSeconds);
    }

    /**
     * 转换为数字（像Excel日期值）
     */
    valueOf() {
        if (this.Year != null && this.Year > 1900) {
            let dt = new Date(this.Year, this.Month - 1, this.Day, this.Hour, this.Minute, this.Second);
            let start = new Date(1900, 0, 1);
            let diffTime = dt.getTime() - start.getTime();
            let days = Math.floor(diffTime / (1000 * 60 * 60 * 24) + 2);
            days += (this.Hour + (this.Minute + this.Second / 60) / 60) / 24;
            return days;
        }
        return (this.Day || 0) + (this.Hour + (this.Minute + this.Second / 60) / 60) / 24;
    }

    /**
     * 转换为长整型数字，格式：年月日时分秒
     */
    ToLong() {
        let d = 0;
        if (this.Year != null) {
            d += this.Year * 10000000000;
        }
        if (this.Month != null) {
            d += this.Month * 100000000;
        }
        if (this.Day != null) {
            d += this.Day * 1000000;
        }
        d += this.Hour * 10000;
        d += this.Minute * 100;
        d += this.Second;
        return d;
    }

    /**
     * 静态方法：解析字符串为MyDate对象
     */
    static Parse(txt) {
        let t = txt.trim();
        
        // 匹配年月日 时分秒格式
        let dateTimeRegex = /^(\d{4})[-/](\d{1,2})[-/](\d{1,2})\s+(\d{1,2}):(\d{1,2}):(\d{1,2})$/;
        let match = t.match(dateTimeRegex);
        if (match) {
            return new MyDate(
                parseInt(match[1]),
                parseInt(match[2]),
                parseInt(match[3]),
                parseInt(match[4]),
                parseInt(match[5]),
                parseInt(match[6])
            );
        }
        
        // 匹配年月日 时分格式
        let dateTimeRegex2 = /^(\d{4})[-/](\d{1,2})[-/](\d{1,2})\s+(\d{1,2}):(\d{1,2})$/;
        match = t.match(dateTimeRegex2);
        if (match) {
            return new MyDate(
                parseInt(match[1]),
                parseInt(match[2]),
                parseInt(match[3]),
                parseInt(match[4]),
                parseInt(match[5]),
                0
            );
        }
        
        // 匹配年月日格式
        let dateRegex = /^(\d{4})[-/](\d{1,2})[-/](\d{1,2})$/;
        match = t.match(dateRegex);
        if (match) {
            return new MyDate(
                parseInt(match[1]),
                parseInt(match[2]),
                parseInt(match[3]),
                0,
                0,
                0
            );
        }
        
        // 匹配时分秒格式
        let timeRegex = /^(\d{1,2}):(\d{1,2}):(\d{1,2})$/;
        match = t.match(timeRegex);
        if (match) {
            return new MyDate(
                null,
                null,
                null,
                parseInt(match[1]),
                parseInt(match[2]),
                parseInt(match[3])
            );
        }
        
        // 匹配时分格式
        let timeRegex2 = /^(\d{1,2}):(\d{1,2})$/;
        match = t.match(timeRegex2);
        if (match) {
            return new MyDate(
                null,
                null,
                null,
                parseInt(match[1]),
                parseInt(match[2]),
                0
            );
        }
        
        return null;
    }
}

// 浏览器支持
if (typeof window !== 'undefined') {
    window.MyDate = MyDate;
}

File: csharp/ToolGood.Algorithm/Operands/MyDate.cs
using System;
using System.Globalization;
using System.Text;
using ToolGood.Algorithm.Internals;

namespace ToolGood.Algorithm.Operands
{
    /// <summary>
    /// MyDate
    /// </summary>
    public sealed class MyDate
    {
        /// <summary>
        /// 年
        /// </summary>
        public int? Year { get; private set; }

        /// <summary>
        /// 月
        /// </summary>
        public int? Month { get; private set; }

        /// <summary>
        /// 日
        /// </summary>
        public int? Day { get; private set; }

        /// <summary>
        /// 时
        /// </summary>
        public int Hour { get; private set; }

        /// <summary>
        /// 分
        /// </summary>
        public int Minute { get; private set; }

        /// <summary>
        /// 秒
        /// </summary>
        public int Second { get; private set; }

        private MyDate()
        { }

        /// <summary>
        /// 构造函数
        /// </summary>
        /// <param name="year">年</param>
        /// <param name="month">月</param>
        /// <param name="day">日</param>
        /// <param name="hour">时</param>
        /// <param name="minute">分</param>
        /// <param name="second">秒</param>
        public MyDate(int? year, int? month, int? day, int hour, int minute, int second)
        {
            Year = year;
            Month = month;
            Day = day;
            Hour = hour;
            Minute = minute;
            Second = second;
        }

        /// <summary>
        /// 构造函数
        /// </summary>
        /// <param name="dt">日期时间</param>
        public MyDate(DateTime dt)
        {
            Year = dt.Year;
            Month = dt.Month;
            Day = dt.Day;
            Hour = dt.Hour;
            Minute = dt.Minute;
            Second = dt.Second;
        }

        /// <summary>
        /// 构造函数
        /// </summary>
        /// <param name="dt">时间</param>
        public MyDate(TimeSpan dt)
        {
            Year = null;
            Month = null;
            Day = dt.Days;
            Hour = dt.Hours;
            Minute = dt.Minutes;
            Second = dt.Seconds;
        }

        /// <summary>
        /// 字符串转MyDate
        /// </summary>
        /// <param name="txt"></param>
        /// <returns></returns>
        public static MyDate Parse(string txt)
        {
            var cultureInfo = CultureInfo.InvariantCulture;// CultureInfo.GetCultureInfo("zh-cn");
            var t = txt.Trim();
            var m = RegexHelper. dateTimeRegex.Match(t); // 年月日 时分秒
            if (m.Success == false) { m = RegexHelper.dateTimeRegex2.Match(t); }
            if (m.Success) {
                var date = new MyDate {
                    Year = int.Parse(m.Groups[(1)].Value, cultureInfo),
                    Month = int.Parse(m.Groups[(2)].Value, cultureInfo),
                    Day = int.Parse(m.Groups[(3)].Value, cultureInfo),
                    Hour = int.Parse(m.Groups[(4)].Value, cultureInfo),
                    Minute = int.Parse(m.Groups[(5)].Value, cultureInfo),
                    Second = int.Parse(m.Groups[(6)].Value, cultureInfo)
                };
                return date;
            }

            m = RegexHelper.dateTimeRegex3.Match(t);// 年月日 时分
            if (m.Success == false) { m = RegexHelper.dateTimeRegex4.Match(t); }
            if (m.Success) {
                var date = new MyDate {
                    Year = int.Parse(m.Groups[(1)].Value, cultureInfo),
                    Month = int.Parse(m.Groups[(2)].Value, cultureInfo),
                    Day = int.Parse(m.Groups[(3)].Value, cultureInfo),
                    Hour = int.Parse(m.Groups[(4)].Value, cultureInfo),
                    Minute = int.Parse(m.Groups[(5)].Value, cultureInfo)
                };
                return date;
            }
            m = RegexHelper.dateRegex.Match(t);// 年月日
            if (m.Success == false) { m = RegexHelper.dateRegex2.Match(t); }
            if (m.Success) {
                var date = new MyDate {
                    Year = int.Parse(m.Groups[(1)].Value, cultureInfo),
                    Month = int.Parse(m.Groups[(2)].Value, cultureInfo),
                    Day = int.Parse(m.Groups[(3)].Value, cultureInfo)
                };
                return date;
            }

            m = RegexHelper.dayTimeRegex.Match(t);// 日 时分秒
            if (m.Success) {
                var date = new MyDate {
                    Day = int.Parse(m.Groups[(1)].Value, cultureInfo),
                    Hour = int.Parse(m.Groups[(2)].Value, cultureInfo),
                    Minute = int.Parse(m.Groups[(3)].Value, cultureInfo),
                    Second = int.Parse(m.Groups[(4)].Value, cultureInfo)
                };
                return date;
            }
            m = RegexHelper.dayTimeRegex2.Match(t);// 日 时分
            if (m.Success) {
                var date = new MyDate {
                    Day = int.Parse(m.Groups[(1)].Value, cultureInfo),
                    Hour = int.Parse(m.Groups[(2)].Value, cultureInfo),
                    Minute = int.Parse(m.Groups[(3)].Value, cultureInfo),
                };
                return date;
            }

            m = RegexHelper.timeRegex.Match(t);// 时分秒
            if (m.Success) {
                var date = new MyDate {
                    Hour = int.Parse(m.Groups[(1)].Value, cultureInfo),
                    Minute = int.Parse(m.Groups[(2)].Value, cultureInfo),
                    Second = int.Parse(m.Groups[(3)].Value, cultureInfo)
                };
                return date;
            }
            m = RegexHelper.timeRegex2.Match(t);// 时分
            if (m.Success) {
                var date = new MyDate {
                    Hour = int.Parse(m.Groups[(1)].Value, cultureInfo),
                    Minute = int.Parse(m.Groups[(2)].Value, cultureInfo)
                };
                return date;
            }
            return null;
        }

        /// <summary>
        /// ToString
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            var stringBuffer = new StringBuilder();
            if (Year != null && Year > 0) {
                stringBuffer.Append(Year);
                stringBuffer.Append('-');
                stringBuffer.Append(Month.Value.ToString("D2"));
                stringBuffer.Append('-');
                stringBuffer.Append(Day.Value.ToString("D2"));

                stringBuffer.Append(' ');
                stringBuffer.Append(Hour.ToString("D2"));
                stringBuffer.Append(':');
                stringBuffer.Append(Minute.ToString("D2"));
                stringBuffer.Append(':');
                stringBuffer.Append(Second.ToString("D2"));
            } else {
                if (Day != null && Day > 0) {
                    stringBuffer.Append(Day);
                    stringBuffer.Append(' ');
                }
                stringBuffer.Append(Hour.ToString("D2"));
                stringBuffer.Append(':');
                stringBuffer.Append(Minute.ToString("D2"));
                stringBuffer.Append(':');
                stringBuffer.Append(Second.ToString("D2"));
            }
            return stringBuffer.ToString();
        }

        internal string ToString(string f)
        {
            if (Year == null || Year == 0) {
                return this.ToString(f);
            }
            return this.ToDateTime().ToString(f);
        }

        /// <summary>
        /// 转DateTime
        /// </summary>
        /// <returns></returns>
        public DateTime ToDateTime()
        {
            return new DateTime(Year ?? 0, Month ?? 0, Day ?? 0, Hour, Minute, Second, DateTimeKind.Utc);
        }

        /// <summary>
        /// 转DateTime
        /// </summary>
        /// <returns></returns>
        public DateTime ToDateTime(DateTimeKind dateTimeKind)
        {
            return new DateTime(Year ?? 0, Month ?? 0, Day ?? 0, Hour, Minute, Second, dateTimeKind);
        }

        /// <summary>
        /// 转TimeSpan
        /// </summary>
        /// <returns></returns>
        public TimeSpan ToTimeSpan()
        {
            return new TimeSpan(Day ?? 0, Hour, Minute, Second);
        }

        /// <summary>
        /// 添加年
        /// </summary>
        /// <param name="year"></param>
        /// <returns></returns>
        public MyDate AddYears(int year)
        {
            var t = (this.Year ?? 0) + year;
            return new MyDate(t, Month, Day, Hour, Minute, Second);
            //return new MyDate(ToDateTime().AddYears(year));
        }

        /// <summary>
        /// 添加月
        /// </summary>
        /// <param name="month"></param>
        /// <returns></returns>
        public MyDate AddMonths(int month)
        {
            var t = (this.Month ?? 0) + month;
            if (t >= 1 && t <= 12) {
                return new MyDate(Year, t, Day, Hour, Minute, Second);
            }
            return new MyDate(ToDateTime().AddMonths(month));
        }

        /// <summary>
        /// 添加日
        /// </summary>
        /// <param name="day"></param>
        /// <returns></returns>
        public MyDate AddDays(int day)
        {
            if (Year != null && Year > 1900) {
                var d = (this.Day ?? 0) + day;
                if (d >= 1 && d <= 28) {
                    return new MyDate(Year, Month, d, Hour, Minute, Second);
                }
                return new MyDate(ToDateTime().AddDays(day));
            }
            return new MyDate(ToTimeSpan().Add(new TimeSpan(day, 0, 0, 0)));
        }

        /// <summary>
        /// 添加时
        /// </summary>
        /// <param name="hour"></param>
        /// <returns></returns>
        public MyDate AddHours(int hour)
        {
            var t = this.Hour + hour;
            if (t >= 0 && t < 24) {
                return new MyDate(Year, Month, Day, t, Minute, Second);
            }
            if (Year != null && Year > 1900) {
                return new MyDate(ToDateTime().AddHours(hour));
            }
            return new MyDate(ToTimeSpan().Add(new TimeSpan(hour, 0, 0)));
        }

        /// <summary>
        /// 添加分
        /// </summary>
        /// <param name="minute"></param>
        /// <returns></returns>
        public MyDate AddMinutes(int minute)
        {
            var t = this.Minute + minute;
            if (t >= 0 && t <= 59) {
                return new MyDate(Year, Month, Day, Hour, t, Second);
            }
            if (Year != null && Year > 1900) {
                return new MyDate(ToDateTime().AddMinutes(minute));
            }
            return new MyDate(ToTimeSpan().Add(new TimeSpan(0, minute, 0)));
        }

        /// <summary>
        /// 添加秒
        /// </summary>
        /// <param name="second"></param>
        /// <returns></returns>
        public MyDate AddSeconds(int second)
        {
            var t = this.Second + second;
            if (t >= 0 && t <= 59) {
                return new MyDate(Year, Month, Day, Hour, Minute, t);
            }
            if (Year != null && Year > 1900) {
                return new MyDate(ToDateTime().AddSeconds(second));
            }
            return new MyDate(ToTimeSpan().Add(new TimeSpan(0, 0, second)));
        }

        #region operator

        /// <summary>
        /// DateTime=>MyDate
        /// </summary>
        /// <param name="myDate"></param>
        public static implicit operator MyDate(DateTime myDate)
        {
            return new MyDate(myDate);
        }

        /// <summary>
        /// MyDate=>DateTime
        /// </summary>
        /// <param name="myDate"></param>
        public static implicit operator DateTime(MyDate myDate)
        {
            return myDate.ToDateTime();
        }

        /// <summary>
        /// MyDate=>TimeSpan
        /// </summary>
        /// <param name="myDate"></param>
        public static implicit operator TimeSpan(MyDate myDate)
        {
            return myDate.ToTimeSpan();
        }

        #endregion operator

        internal long ToLong()
        {
            long d = 0;
            if (Year != null) {
                d += Year.Value * 10000000000;
            }
            if (Month != null) {
                d += Month.Value * 100000000;
            }
            if (Day != null) {
                d += Day.Value * 1000000;
            }
            d += Hour * 10000;
            d += Minute * 100;
            d += Second;
            return d;
		}


	}
}